sourcegraph/sourcegraph
# search: Restructure "sync query from URL" logic (2nd try)

[The original PR](https://github.com/sourcegraph/sourcegraph/pull/41459) got reverted because it cleared the query input on
repo pages when clicking certain links. That's because processing such a
URL would return an empty query, which was then synced to the query
state.

ec694d7a2c3ec57523815dd96e2bceb1401e8e5b contains code from the original PR (does not need to be reviewed)
[`7d544d9aff3fda337d1a8d9724c41544e9dc14b3`](https://github.com/sourcegraph/sourcegraph/pull/41550/commits/7d544d9aff3fda337d1a8d9724c41544e9dc14b3) contains the fix.

The original code (before the refactor) treated non-existing queries as
"empty" and simply didn't update state for empty quries.
While I think the more correct approach is to only sync query information
from the URL when the URL actually contains query parameters, I don't
want to introduce too many changes at this point. So the additional changes
do not update the query state when the query is empty.

I added a new integration test and verified that it fails before the
fix and is works after applying the fix.

## Test plan

Integration test

Manual testing:

- Open any file: The query input should show repo and file
- Select line(s) in file: The query input should still show repo and file
  (the input was cleared before the fix)

## App preview:

- [Web](https://sg-web-fkling-querystate-p2.onrender.com)
- [Storybook](https://5f0f381c0e50750022dc6bf7-eneqtpbaxu.chromatic.com)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.
